torch/_refs/nn/functional/__init__.py

```diff
@@ -577,3 +579,30 @@ def glu(a: TensorLikeType, dim: int = -1) -> TensorLikeType:
     b, c = torch.tensor_split(a, 2, dim)
 
     return b * torch.sigmoid(c)
+
+
+@register_decomposition(torch.ops.aten.pairwise_distance)
+@out_wrapper()
+def pairwise_distance(
+    x1: TensorLikeType,
+    x2: TensorLikeType,
+    p: NumberType = 2.0,
+    eps: NumberType = 1e-6,
+    keepdim=False,
+) -> TensorLikeType:
+    return torch.linalg.vector_norm(x1 - x2 + eps, ord=p, dim=-1, keepdim=keepdim)
+
+
+@register_decomposition(torch.ops.aten.pdist)
+@elementwise_type_promotion_wrapper(
+    type_promoting_args=("a",),
+    type_promotion_kind=utils.ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
+)
+@out_wrapper()
+def pdist(a: TensorLikeType, p: int = 2) -> TensorLikeType:
+    check(a.ndim == 2, lambda: f"pdist only supports 2D tensors, got: {a.ndim}D")
+    check(p >= 0, lambda: "pdist only supports non-negative p values")
+    t = torch.linalg.vector_norm(a.unsqueeze(1) - a, ord=p, dim=2)
```

Comment:
this decomposition is quite catastrophic without compiler (even more so than usual decompositions), increasing memory use by orders of maginute, but even with compiler it's unclear if compiler would be able to recover matmul that's typically used for efficient implementation of cdist/pdist

Reply:
Hadn't thought of that, that is indeed terrible.

For `p=2` you can write this using a `matmul` with the transpose like you say (and maybe it is worth it to have that in as a special case), but for generic p I don't think you can.

I am not seeing a way of making this more efficient for generic p. You need to compute

$$ D_{ij} = \sum_a \left| a_{ia} - a_{ja}\right|^p $$

and for $p=2$ you can drop the absolute value and expand to write it in terms of `a @ a.T` but for generic `p` I can't see a way of doing without creating the intermediary tensor $M_{ija}=a_{ia} - a_{ja}$ or doing some explicit for loops. Any suggestions?

Reply:
YEs, you are right, matmul solution is applicable to p=2 only (but also, let's face it, 99% of pdist calls in the wild are with p=2). It's fine to have a special case for p=2, and use your decomposition for everything else (because that's right, there's no obvious better way).

Reply:
I added this. One thought: should I document somewhere that this is expected to be inefficient for `p!=2`? I will add a comment in the code, but wondering if there is some other place to note that in.

Reply:
I think it's fine really, as this is expected if you put some thought into it.